normalizer_lib.py

```diff
@@ -1,205 +1,224 @@
+            if difference(acceptor(token), valid_token).num_states() == 0:
+                returned.append(token)
+            else:
+                returned.append("<UNK>")
+        return " ".join(returned)
+
+
+    def pass_only_valid_sentences(self, string) -> List[str]:
+        """Replaces invalid sentences.
+
+        Args:
+            string: A line from a corpus.
+
+        Returns:
+            The line or the line replaced with <SENTENCE_REJECTED>.
+        """
+        valid_token = (self.initial_punctuation.ques +
+                       self.verbalizable() +
+                       self.final_punctuation.ques)
+        remove_extra_whitespace = (string @ self.remove_extra_whitespace()).optimize().string()
+        split_string = remove_extra_whitespace.split(" ")
+        for token in split_string:
+            if difference(acceptor(token), valid_token).num_states() != 0:
+                return "<SENTENCE_REJECTED>"
+        return string
+
+
+    # Detach punctuation from words
+    # e.g. "Who are you?" -> "Who are you ?"
+
+
+    def detach_punctuation(self) -> Fst:
+        "Detaches punctuation from words."
+        non_grapheme_punct = difference(self.punctuation, self.graphemes)
+        detach_leading_punctuation = cdrewrite(
+            pynutil.insert(" "),
+            union("[BOS]", byte.SPACE) + non_grapheme_punct.plus,
+            union(non_grapheme_punct, self.graphemes, self.numerals),
+            self.SIGMA_STAR)
+        detach_trailing_punctuation = cdrewrite(
+            pynutil.insert(" "),
+            union(non_grapheme_punct, self.graphemes, self.numerals, "/"),
+            self.punctuation.plus + union("[EOS]", byte.SPACE),
+            self.SIGMA_STAR)
+        return (detach_leading_punctuation @ detach_trailing_punctuation).optimize()
+
+
+    # Delete freestanding punctuation
+    # e.g. "hi there ?" -> "hi there
+
+
+    def delete_freestanding_punctuation(self) -> Fst:
+        "Deletes freestanding punctuation."
+        delete_freestanding_punctuation = cdrewrite(
+            pynutil.delete(self.punctuation),
+            union("[BOS]", byte.SPACE),
+            union("[EOS]", byte.SPACE),
+            self.SIGMA_STAR)
+        return delete_freestanding_punctuation
+
+
+    def apply_fst_rules(self, string: str) -> str:
+        """Applies FST rewrite rules."""
+        return string if (string == "<STRING_REJECTED>") else (
+            string
+            @ self.language_specific_fixes()
+            @ self.detach_punctuation()
+            @ self.delete_freestanding_punctuation()
+            @ self.remove_extra_whitespace()
+            ).optimize().string()
+
+
+    def token_normalizer(self, string: str) -> str:
+        """Normalizes text by applying FST rewrite rules.
+
+        Args:
+            string: A line from a corpus.
+
+        Returns:
+            The normalized line from the corpus.
+        """
+        string = unicodedata.normalize("NFC", string.lower())
```

Comment:
Note that Python's lower() method might not work for all scripts, it's probably okay for our purposes but maybe note it in the docstring.

Reply:
Added to the docstring for the new `preprocess_string` helper method where `.lower()` is called.